keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -51,223 +48,183 @@
-                  volume plugin reports an error and kubelet fails to start the pod.
-                  It is CSI driver fault that it advertises SELinux support and then fails to apply it.
-
-* Nothing changes when `CSIDriver.Spec.SELinuxMountSupported` is `false` or not set:
-    * CSI volume plugin calls CSI without any special SELinux mount options and it autodetects, if the volume supports SELinux or not by presence of `seclabel` mount option.
-      This is current kubelet behavior.
-
-* Nothing changes if kubelet does not know the SELinux context of a pod (`SELinuxOptions.Level` is empty) or pod's `SELinuxRelabelPolicy` is `Always`.
-    * Volume is mounted without any SELinux options and passed to the container runtime with or without ":Z", depending on if the volume plugin supports SELinux or not (by checking `seclabel` mount option).
-      The container runtime allocates a new SELinux context and recursively relabels all files on the volume.
-      This is current kubelet behavior.
-
-Validation:
+#### Behavioral changes
+This KEP changes behavior of Kubernetes when two pods with different SELinux contexts use the same volume.
+Let Pod A with SELinux context X runs and Pod B with SELinux context Y is about to start on the same node and both use the same volume.
 
-* Kubernetes checks that `SELinuxRelabelPolicy` field can be used in a pod only when at least `SELinuxOptions.Level` is set.
+* *Before this KEP*: Pod A suddenly starts getting "permission denied" errors when accessing files on the volume, because the container runtime re-labeled all files on it with label Y when starting pod B. Pod B will start just fine and can access the volume.
+* *As proposed in this KEP*: Pod B won't even start, because the volume is already mounted with `-o context=X`. When kubelet tries to mount the same volume with `-o context=Y`, this mount fails. The Pod B with be `ContainerCreating` until Pod A is deleted and its volumes unmounted.
 
-When a Pod specifies incomplete SELinux label (i.e. omits `SELinuxOptions.User`, `.Role` or `.Type`), kubelet fills the blanks from the system defaults provided by [ContainerLabels() from go-selinux bindings](https://github.com/opencontainers/selinux/blob/621ca21a5218df44259bf5d7a6ee70d720a661d5/go-selinux/selinux_linux.go#L770).
+A special case of the previous example is when two pods with different SELinux contexts use the same volume, but different subpaths of it.
+The container runtime then re-labels only these subpaths and as long as the subpaths are different, both pods can run today.
+**This will not be possible with this KEP** - while the container runtime operates on subpaths, kubelet always mounts the whole volume. The first mount with `-o context=X` will succeed, the second with `-o context=Y` will fail. The second Pod will be `ContainerCreating` as described above.
```

Comment:
Will we have events/messages indicating why Pod B stays in ContainerCreating in this case?

Reply:
Same as above, mount will fail, most probably with a very generic message.

Reply:
Added a better message above when kubelet knows that the SELinux contexts are wrong.